### PR TITLE
Implement persona-hub using Curator

### PR DESCRIPTION
**Pull request by madiator**
Source: https://github.com/tencent-ailab/persona-hub/blob/main/code/openai_synthesize.py

It can be seen that their code uses a for loop to prompt the LLM per persona. 
The npc type of generation takes about 30 seconds per example. So 200k personas should take 69 days.
If someone is at tier 5, we can get a speedup of 15k (since we can send 30k requests per minute in tier 5 whereas without curator it is 2 requests per minute). So that brings down to a modest 6.6 minutes.

### CODE REVIEW

File: examples/persona-hub/synthesize.py [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+        temperature=0.7,
+    )
+    return generator
+
+
+def main(args):
+    template = get_template(args.template)
+    generator = get_generator(template)
+    # Load the persona dataset
+    persona_dataset = load_dataset("proj-persona/PersonaHub", data_files="persona.jsonl")["train"]
```
Comment (RyanMarten) [inline examples/persona-hub/synthesize.py:42]:
```suggestion
    persona_dataset = load_dataset("proj-persona/PersonaHub", data_files="persona.jsonl", split="train")
```

File: examples/persona-hub/synthesize.py [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+    if args.sample_size > 0:
+        persona_dataset = persona_dataset.select(range(args.sample_size))
```
Comment (RyanMarten) [inline examples/persona-hub/synthesize.py:44]:
```suggestion
        persona_dataset = persona_dataset.take(args.sample_size)
```

File: examples/persona-hub/synthesize.py [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+        return prompt_templates.math_template
+    else:
+        raise ValueError(
+            "Invalid template type. Choose from 'instruction', 'knowledge', 'npc', or 'math'."
+        )
+
+
+def get_generator(template):
+    def prompt_func(row):
+        return template.format(persona=row["persona"])
```
Comment (RyanMarten) [inline examples/persona-hub/synthesize.py:28]: Nice, I like this

**Review by RyanMarten — COMMENTED** [review]
Ran into errors since my key didn't have gpt-4o access

We should catch that all of the requests failed and raise a relevant error instead of having the user exposed to the responses file

```
d with Exception API error: {'message': 'Project `proj_849B36OelpiRtrG7hEPPgVSD` does not have access to model `gpt-4o`', 'type': 'invalid_request_error', 'param': None, 'code': 'model_not_found'}, attempts left 4
2024-12-04 18:44:30,733 - bespokelabs.curator.request_processor.base_online_request_processor - INFO - Adding request 0 to retry queue. Will retry in next available slot. Attempts remaining: 3
Processing OpenAIOnlineRequestProcessor requests:   0%|                                           | 0/10 [00:00<?, ?it/s]
2024-12-04 18:44:30,736 - bespokelabs.curator.request_processor.base_online_request_processor - INFO - Processing complete. Results saved to /Users/ryan/.cache/curator/f467067fea2470e9/responses_0.jsonl
2024-12-04 18:44:30,737 - bespokelabs.curator.request_processor.base_online_request_processor - INFO - Status tracker: Tasks - Started: 10, In Progress: 10, Succeeded: 0, Failed: 0, Already Completed: 0
Errors - API: 20, Rate Limit: 0, Other: 20, Total: 40
Traceback (most recent call last):
  File "/Users/ryan/curator/examples/persona-hub/synthesize.py", line 75, in <module>
    main(args)
  File "/Users/ryan/curator/examples/persona-hub/synthesize.py", line 46, in main
    output = generator(persona_dataset)
             ^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/ryan/curator/src/bespokelabs/curator/prompter/prompter.py", line 191, in __call__
    return self._completions(self._request_processor, dataset, working_dir)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/ryan/curator/src/bespokelabs/curator/prompter/prompter.py", line 279, in _completions
    dataset = request_processor.run(
              ^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/ryan/curator/src/bespokelabs/curator/request_processor/base_online_request_processor.py", line 177, in run
    return self.create_dataset_files(working_dir, parse_func_hash, prompt_formatter)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/ryan/curator/src/bespokelabs/curator/request_processor/base_request_processor.py", line 200, in create_dataset_files
    raise ValueError(f"No responses files found in {working_dir}")
ValueError: No responses files found in /Users/ryan/.cache/curator/f467067fea2470e9
```

**Review by RyanMarten — APPROVED** [review]
Re-ran it with a valid API key and successfully generated the file

File: examples/persona-hub/synthesize.py (hunk 1 above)
Comment (madiator) [reply examples/persona-hub/synthesize.py]: Done

File: examples/persona-hub/synthesize.py (hunk 2 above)
Comment (madiator) [reply examples/persona-hub/synthesize.py]: Done

**Pull request merged**